libs/@blockprotocol/graph/src/types/subgraph/edges.ts

```diff
@@ -9,18 +10,17 @@ export * from "./edges/kind.js";
 export * from "./edges/outward-edge.js";
 export * from "./edges/variants.js";
 
-/** @todo - Re-express these and `Vertices` as `Record`s? */
+export type OntologyRootedEdges = Record<
+  BaseUri,
+  Record<OntologyTypeRevisionId, OntologyOutwardEdge[]>
+>;
 
-export type OntologyRootedEdges = {
-  [typeBaseUri: BaseUri]: {
-    [typeVersion: number]: OntologyOutwardEdge[];
-  };
-};
+export type KnowledgeGraphRootedEdges = Record<
+  EntityId,
+  Record<Timestamp, KnowledgeGraphOutwardEdge[]>
+>;
 
-export type KnowledgeGraphRootedEdges = {
-  [entityId: EntityId]: {
-    [edgeFirstCreatedAt: Timestamp]: KnowledgeGraphOutwardEdge[];
-  };
-};
-
-export type Edges = OntologyRootedEdges & KnowledgeGraphRootedEdges;
+// We technically want to intersect (`&`) the types here, but as their property keys overlap it confuses things and we
+// end up with unsatisfiable values like `EntityVertex & DataTypeVertex`. While the union (`|`) is semantically
+// incorrect, it structurally matches the types we want.
+export type Edges = OntologyRootedEdges | KnowledgeGraphRootedEdges;
```

Comment:
This is really frustrating. If we use `&` as we'd like to, then TS thinks the keys overlap and produces an unsatisfiable type
![image](https://user-images.githubusercontent.com/25749103/216955015-e745bcce-5fca-44cb-8500-7b4e37917b45.png)

As they’re index signature property thingymajigsm it treats them as if they were a same key and then distributes the `&` all the way down through the `string` properties. Generally this distributive property works out (if you actually know the keys) for example in this where it combines the two `y` properties but leaves `x` and `z` as they were

<img src="https://user-images.githubusercontent.com/25749103/216955202-5ee494c1-1e50-4b3e-b78f-4dd3f25f9602.png" width="300" />

But because of the index signature types it just assumes that the keys are the same and combines the constraints.

TypeScript's structural typing does mean that `x: Foo` is assignable to `Foo | Bar` and `x: Bar` is assignable to `Foo | Bar`, _and importantly_ `x: Foo & Bar` is assignable to `Foo | Bar`, so `OntologyRootedEdges | KnowledgeGraphRootedEdges` works fine for us as a type.

Reply:
Does this have any implicatiosn for the consumers of `Edges`?

Reply:
Yes, but the only way I feel like I can answer that is by saying look at the rest of the diff of the PR :( 

It's all sort of weird obscure stuff to do with how well TS can infer/constrain types while traversing it.